### ISSUE TITLE
Incomplete clean-up in versioned bucket

### ISSUE THREAD

**Issue by DomDerrien** [onhate/cdk-gc#4]
Tl;dr: in a bucket with _versioning_ enabled, the script creates `delete markers` ending up using more space than before...

I run the `cdk-gc` script b/c my CDK bucket was storing about ~2,700 files.
The next morning, S3 Lens reported that the file numbers grew up to 5.400 files, even if the S3 console was only showing `570` as the number of contained assets. I ran two series of AWS CLI commands:
- `aws s3api list-objects-v2 --bucket ... --profile ...` => ~570 entries in the `Contents` array.
- `aws s3api list-object-versions --bucket ... --profile ...` => ~5,400 entries: ~2,700 entries in the `Versions` array and ~2,700 entries in the `DeletedMarkers` array.

In the call to [s3.deleteObject()](https://github.com/onhate/cdk-gc/blob/main/src/gc.js#L184), the version identifier of the file to be permanently deleted is missing!

Hopefully, I tried the script in my `dev` account and will be able to clean everything with the second extract.

Ideally, the script should check the _versioning_ state of the bucket before deleting the ZIP files, and process with the deletion more carefully. As the files are named with the hash compute from their content, the CDK does not re-upload the ZIP files twice. So the each file history contains only one version, and the script will trigger only one _delete_ operation per file.

Let me know if you have a fix in mind or if you're open to a PR.

--
Reference: [AWS SDK for JavaScript v3](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest) > [@aws-sdk/client-s3](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-client-s3) > [Client](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/s3) > [DeleteObjectCommand](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/s3/command/DeleteObjectCommand/)

**Comment by onhate** [comment]
I did not know it was possible to configure CDK to have a versioned bucket, is it something CDK did or you changed the bucket manually?

**Comment by onhate** [comment]
Ow, I see, by default it is versioned, mine is too. Hmmm... let me try a fix for that

**Comment by onhate** [comment]
I can send the versionId in the delete operation, but to solve your problem you may want to create a Lifecycle Rule to cleanup the other versions, WDYT?

**Comment by onhate** [comment]
This may help you with the cleanup while I fix this.

```bash
#!/bin/bash -e

PROFILE="XXXXXX"
BUCKET_NAME="cdk-hnb659fds-assets-00000000000-sa-east-1"

function main {
	local IFS=
	local versionKeyRegexp=$'^([^\t]+)\t(.+)'
	while read -r item; do
		if [[ $item =~ $versionKeyRegexp ]]; then
			local versionId=${BASH_REMATCH[1]}
			local key=${BASH_REMATCH[2]}

			aws s3api delete-object \
				--bucket "$BUCKET_NAME" \
				--key "$key" \
				--version-id "$versionId" \
				--no-cli-pager \
				--profile "$PROFILE"

			echo "Deleted: [$key] | [$versionId]"
		fi
	done < <(aws s3api list-object-versions \
		--bucket "$BUCKET_NAME" \
		--output text \
		--query "[DeleteMarkers,Versions][].{a:VersionId,b:Key}" \
		--no-cli-pager \
		--profile "$PROFILE")
}

main

```

**Comment by onhate** [comment]
Here is a faster version in node.js (aws cli is very slow for multiple commands)


```js
const { S3Client, ListObjectVersionsCommand, DeleteObjectCommand } = require('@aws-sdk/client-s3');

const profile = 'XXXXXX';
const bucket = 'cdk-hnb659fds-assets-000000000000-sa-east-1';
const region = 'sa-east-1';

const s3 = new S3Client({ profile, region });

async function main() {
  const { Versions, IsTruncated } = await s3.send(new ListObjectVersionsCommand({ Bucket: bucket }));

  for (const version of Versions) {
    const deleteParams = {
      Bucket: bucket,
      Key: version.Key,
      VersionId: version.VersionId
    };

    await s3.send(new DeleteObjectCommand(deleteParams));
    console.log(`Deleted: [${version.Key}] | [${version.VersionId}]`);
  }

  if (IsTruncated) {
    console.log('IsTruncated: true');
    return main();
  }
}

main().catch(console.error);
```

```
node cleanup.js
```

**Comment by onhate** [comment]
here is a faster version in node.js 


```js
const { S3Client, ListObjectVersionsCommand, DeleteObjectCommand } = require('@aws-sdk/client-s3');

const profile = 'drozdev';
const bucket = 'cdk-hnb659fds-assets-726342070183-sa-east-1';
const region = 'sa-east-1';

const s3 = new S3Client({ profile, region });

async function main() {
  const { DeleteMarkers, IsTruncated } = await s3.send(new ListObjectVersionsCommand({
    Bucket: bucket
  }));

  for (const version of DeleteMarkers) {
    if (version.IsLatest) {
      const deleteParams = {
        Bucket: bucket,
        Key: version.Key,
        VersionId: version.VersionId
      };

      await s3.send(new DeleteObjectCommand(deleteParams));
      console.log(`Deleted: [${version.Key}] | [${version.VersionId}]`);
    }
  }

  if (IsTruncated) {
    console.log('IsTruncated: true');
    return main();
  }
}

main().catch(console.error);
```


```
node cleanup.js
```

**Comment by DomDerrien** [comment]
> I can send the versionId in the delete operation, but to solve your problem you may want to create a Lifecycle Rule to cleanup the other versions, WDYT?

Sending the `versionId` should solve the issue.

> This may help you with the cleanup while I fix this.

Thanks for the script. However, it's incomplete because both the delete markers and the initial file version have to be deleted.

Don't worry for my situation, I'm going to handle it soon ;)

Thank you for your prompt response.
Keep up with the good work!
